indra/llwindow/llgamecontrol.cpp

```diff
@@ -441,57 +620,144 @@ std::string LLGameControl::Options::AxisOptions::saveToString() const
+    while (true)
     {
-        options.push_back("axis_map:{" + axis_map + "}");
+        size_t pos = source.find_first_of(depth ? "{}" : ",{}", offset);
+        if (pos == std::string::npos)
+        {
+            return !depth && split(source);
+        }
+        if (source[pos] == ',')
+        {
+            if (!split(source.substr(0, pos)))
+                return false;
+            source = source.substr(pos + 1);
+            offset = 0;
+        }
+        else if (source[pos] == '{')
+        {
+            depth++;
+            offset = pos + 1;
+        }
+        else if (depth) // Assume '}' here
+        {
+            depth--;
+            offset = pos + 1;
+        }
+        else
+        {
+            return false; // Extra '}' found
+        }
     }
 
-    std::string button_map = LLStringOps::join<std::map<U8, U8>, std::pair<U8, U8>>(mButtonMap, pair2str);
-    if (!button_map.empty())
+    return true;
+}
+
+void LLGameControl::Options::AxisOptions::loadFromString(std::string options)
+{
+    resetToDefaults();
+
+    if (options.empty())
+        return;
+
+    std::map<std::string, std::string> pairs;
+    if (!parse(pairs, options))
     {
-        options.push_back("button_map:{" + button_map + "}");
+        LL_WARNS("SDL2") << "Invalid axis options: '" << options << "'" << LL_ENDL;
     }
 
-    auto opts2str = [](size_t i, const AxisOptions& options) -> std::string
+    std::string invert = pairs["invert"];
+    if (!invert.empty())
     {
-        std::string string = options.saveToString();
-        return string.empty() ? string : llformat("%u:%s", i, string.c_str());
-    };
+        if (invert != "1")
+        {
+            LL_WARNS("SDL2") << "Invalid invert value: '" << invert << "'" << LL_ENDL;
+        }
+        else
+        {
+            mInvert = true;
+        }
+    }
 
-    std::string axis_options = LLStringOps::join<std::vector<AxisOptions>, AxisOptions>(mAxisOptions, opts2str);
-    if (!axis_options.empty())
+    std::string dead_zone = pairs["dead_zone"];
+    if (!dead_zone.empty())
     {
-        options.push_back("axis_options:{" + axis_options + "}");
+        size_t number = std::stoull(dead_zone);
+        if (number > MAX_AXIS_DEAD_ZONE || std::to_string(number) != dead_zone)
+        {
+            LL_WARNS("SDL2") << "Invalid dead_zone value: '" << dead_zone << "'" << LL_ENDL;
+        }
+        else
+        {
+            mDeadZone = (U8)number;
+        }
     }
 
-    std::string result = LLStringOps::join(options);
-
-    return result.empty() ? result : "{" + result + "}";
+    std::string offset = pairs["offset"];
+    if (!offset.empty())
+    {
+        S32 number = std::stoi(offset);
+        if (abs(number) > MAX_AXIS_OFFSET || std::to_string(number) != offset)
+        {
+            LL_WARNS("SDL2") << "Invalid offset value: '" << offset << "'" << LL_ENDL;
+        }
+        else
+        {
+            mOffset = (U8)number;
+        }
+    }
```

Comment:
`offset` should be allowed to be negative.  The logic will have to clamp a `max` and a `min`.